File: frontend/src/services/PathwayService.ts
import axios from "axios";

const API_BASE_URL = "https://backend-production-31fe.up.railway.app";

export interface Pathway {
  id: number;
  pathway_id: string;
  pathway_name: string;
  category?: string | null;
  super_category?: string | null;
  disease?: string | null;
  disease_stages?: string | null;
  cellular_location?: string | null;
  major_cell_types?: string | null;
  primary_function?: string | null;
  clinical_importance?: string | null;
  druggable?: boolean | null;
  pathway_status?: string | null;
  canonical_databases?: string | null;
  source_sheet: string;
}

export interface PathwayDetails
  extends Pathway {
  pathway_data?: Record<string, unknown> | null;
}

export interface PathwayListResponse {
  items: Pathway[];
  total: number;
  page: number;
  limit: number;
  total_pages: number;
}

const PathwayService = {
  getAll: async (
    page = 1,
    limit = 25
  ): Promise<PathwayListResponse> => {
    const response =
      await axios.get<PathwayListResponse>(
        `${API_BASE_URL}/pathways/`,
        {
          params: {
            page,
            limit,
          },
        }
      );

    return response.data;
  },

  getById: async (
    id: number
  ): Promise<PathwayDetails> => {
    const response =
      await axios.get<PathwayDetails>(
        `${API_BASE_URL}/pathways/${id}`
      );

    return response.data;
  },

  getByCode: async (
    pathwayId: string
  ): Promise<PathwayDetails> => {
    const response =
      await axios.get<PathwayDetails>(
        `${API_BASE_URL}/pathways/code/${encodeURIComponent(
          pathwayId
        )}`
      );

    return response.data;
  },

  search: async (
    keyword: string
  ): Promise<Pathway[]> => {
    const response =
      await axios.get<Pathway[]>(
        `${API_BASE_URL}/pathways/search/${encodeURIComponent(
          keyword
        )}`
      );

    return response.data;
  },

  getCategories: async (): Promise<
    string[]
  > => {
    const response =
      await axios.get<string[]>(
        `${API_BASE_URL}/pathways/categories`
      );

    return response.data;
  },

  getSuperCategories: async (): Promise<
    string[]
  > => {
    const response =
      await axios.get<string[]>(
        `${API_BASE_URL}/pathways/super-categories`
      );

    return response.data;
  },

  getByCategory: async (
    category: string
  ): Promise<Pathway[]> => {
    const response =
      await axios.get<Pathway[]>(
        `${API_BASE_URL}/pathways/category/${encodeURIComponent(
          category
        )}`
      );

    return response.data;
  },
};

export default PathwayService;

File: backend/app/services/pathway_service.py
from math import ceil

from sqlalchemy import or_
from sqlalchemy.orm import Session

from app.models.pathway import Pathway


class PathwayService:

    @staticmethod
    def get_all(
        db: Session,
        page: int = 1,
        limit: int = 25
    ):
        if page < 1:
            page = 1

        if limit < 1:
            limit = 25

        if limit > 100:
            limit = 100

        query = db.query(Pathway)

        total = query.count()

        total_pages = (
            ceil(total / limit)
            if total > 0
            else 0
        )

        offset = (page - 1) * limit

        items = (
            query
            .order_by(Pathway.id)
            .offset(offset)
            .limit(limit)
            .all()
        )

        return {
            "items": items,
            "total": total,
            "page": page,
            "limit": limit,
            "total_pages": total_pages
        }

    @staticmethod
    def get_by_id(
        db: Session,
        pathway_id: int
    ):
        return (
            db.query(Pathway)
            .filter(
                Pathway.id == pathway_id
            )
            .first()
        )

    @staticmethod
    def get_by_pathway_id(
        db: Session,
        pathway_id: str
    ):
        return (
            db.query(Pathway)
            .filter(
                Pathway.pathway_id == pathway_id
            )
            .first()
        )

    @staticmethod
    def get_by_name(
        db: Session,
        pathway_name: str
    ):
        return (
            db.query(Pathway)
            .filter(
                Pathway.pathway_name == pathway_name
            )
            .first()
        )

    @staticmethod
    def search(
        db: Session,
        keyword: str
    ):
        keyword = keyword.strip()

        if not keyword:
            return []

        search_pattern = f"%{keyword}%"

        return (
            db.query(Pathway)
            .filter(
                or_(
                    Pathway.pathway_id.ilike(
                        search_pattern
                    ),
                    Pathway.pathway_name.ilike(
                        search_pattern
                    ),
                    Pathway.category.ilike(
                        search_pattern
                    ),
                    Pathway.super_category.ilike(
                        search_pattern
                    ),
                    Pathway.disease.ilike(
                        search_pattern
                    ),
                    Pathway.primary_function.ilike(
                        search_pattern
                    ),
                    Pathway.major_cell_types.ilike(
                        search_pattern
                    )
                )
            )
            .order_by(
                Pathway.pathway_name
            )
            .all()
        )

    @staticmethod
    def get_categories(
        db: Session
    ):
        results = (
            db.query(
                Pathway.category
            )
            .filter(
                Pathway.category.isnot(None)
            )
            .distinct()
            .order_by(
                Pathway.category
            )
            .all()
        )

        return [
            row[0]
            for row in results
            if row[0]
        ]

    @staticmethod
    def get_super_categories(
        db: Session
    ):
        results = (
            db.query(
                Pathway.super_category
            )
            .filter(
                Pathway.super_category.isnot(None)
            )
            .distinct()
            .order_by(
                Pathway.super_category
            )
            .all()
        )

        return [
            row[0]
            for row in results
            if row[0]
        ]

    @staticmethod
    def get_by_category(
        db: Session,
        category: str
    ):
        return (
            db.query(Pathway)
            .filter(
                Pathway.category == category
            )
            .order_by(
                Pathway.pathway_name
            )
            .all()
        )